### ISSUE TITLE
C++ data object design

### ISSUE THREAD

**Issue by pavlis** [wangyinz/mspass#13]
The core data objects for the C++ elements of mspass are now checked into this repository.   I designed the API for this level of these things to be clean and separable from mspass.   That was intentional to maximize potential reuse of this code, most of which is pretty well exercised and solid.   For the record the core data objects are:
TimeSeries - a scalar seismogram
Seismogram - a three component seismogram (called ThreeComponentSeismogram in seispp of antelope contrib)
Ensemble - a template class for a generic collection of data objects.   Formerly two different parallel classes called TimeSeriesEnsemble and ThreeComponentEnsemble.   With this new design the two old symbols were easily implemented as typedef of the template class.   

My thinking, particularly after a comment you made yesterday, is a next step is to abstract the interface to MongoDB.   An important API decision for both C++ and python is how tightly we want to link data to MongoDB.   A concrete point about that is whether each data class should have a copy of a handle to the db or the db handle code should be a separate procedural library or class that implements what we need.  I'm about 99% sure the later is the only sensible model as I can't conceive how we could keep the handle lightweight enough to pass it around with data unless we reduced it to a raw pointer.  A raw pointer would be evil and a nightmare to maintain.   I would argue that we should just create an abstraction of the MongoDB API that implements only concepts we require for mspass that can interact with mspass data objects.   What is your opinions on this?

Going ahead with this let's list only the concepts we need for the database and not what mongo is capable of doing.   These are largely independent of the decision about whether data should have a pointer to a handle or just call methods in the handle.  Here is a start:

1. What the textbooks and Mongo' documentation call CRUD (Create, Read, Update, Delete).   These have some subtle elements with a secondary decision - do we split Metadata and the seismic data or bundle them in the same "document"?   We discussed this in the proposal and I think you've been doing some early experiments with GridFS.
2." Find" and some generic concept to assure the handle is pointing at one record or a set of common records.  I think Mongo refers to this as a cursor.  In antelope it is the record field of the Dbptr.   I think oracle and postgres use the term cursor too.
3.  "group" - a way to bundle a set of data together with keys.   "group by" in sql or dbgroup in antelope.   May want a group form of a cursor that is an easily traversed list of things that can be used to build an ensemble.  
4.  We need some for of "join" operation.   I think there is, in fact, a join procedure in the mongo api.   How a join behaves in a document db is, however, something I'm not sure I totally get.   Well, I get the idea but not sure I appreciate the potential gotchas.   It is dangerous enough in a relational database if you don't get the keys right.  I've innocently create several million row virtual tables in antelope by using improper join keys.   Suspect this could get worse in mongo with large data sets.  
5.  A global query capability building on antelope's dbquery procedure.  e.g. basic things to ask before starting a program are the number or things matching a find query.   The generic concept we should consider is the need to extract handle attributes.   i.e. we should think of the list of things we support as database attributes themselves and the list could, in fact, be stored on a read only "document" in MongoDB.  That is just and idea and a potential implementation detail.  They key point is a need to query the db for things not related to the data itself, but things related to the handle itself.
6.  Theerror logging capability we discussed a few days ago.   This is really only a CRUD operation, but it might benefit from a simplied API.   e.g. we might create a template like this:
template<typename T> DBHandle::save_error_log(T d); 
This would simplify error handlers and make it easy for a program to log an information message.

Now I suggest the Seismogram and TimeSeries objects should have some variant of the following within mspass:

MPTimeSeries : public ErrorLogger 
MPSeismogram : public Error Logger

ErrorLogger is a class I already checked in with some basic testing.   It allows error handlers to post errors and continue unless and abort only if that is appropriate (e.g. on a malloc failure).   The class has a required error severity that is intended to a guide for error handlers.  It also, however, provides a way to just log something that isn't an error as either a verbose option or just a useful way to log something you aren't sure will be useful.   In research code I've many times used grep to pull lines from some silly program I ran thousands of times to extract the one gem I wanted.  Anyway, the point is the error logger is the primary add on to make a seismogram MP.

The only other things these would probably need are a way to uniquely link the data to the database entry from which they were created or derived.   For Mongo I think this might be as simple as two long ints:
long ObjectId;
long ObjectID_parent;
But I don't know enough about this yet to know if that will be enough state information.  The key concept is the data just have a way to link back to the database.   Could be a pointer to a dbhandle, but that would be REALLY unstable.  Might be feasible with std::unique_ptr, but still not a wise approach in my opinion   

Finally, for the python bindings I would recommend we figure out a way to alias away the MP thing this current design requires.   I think the python wrappers can redefine that easily so an MPSeismogram becomes a Seismogram in the python API.   If that  is too awkward or you think it would be too confusing we could also just rename the C++ Seismogram as CoreSeismogram and let TimeSeries become CoreTimeSeries.   

This got very long - suspect you will need a while to digest it.   Don't hurry as I don't expect to do much on this for the next several days. 

Expecting this could initiate a long new chain

**Comment by wangyinz** [comment]
This is indeed a lot to digest. Let me first comment on the arrangement of the code first. I think it make sense to have the seismic component separated, but I think it might be better to have it within the mspass instead of being parallel to it. I would argue putting them in `include/mspass/seismic` makes more sense, and that is also consistent with the use of mspass namespace in those headers. Otherwise, it is confusing that the directory structure (as well as the cmake file) suggest them belonging to a separate library, but they use the same name space.

Alternatively, since it is apparent that the seismic headers depend on the utility headers, it is probably better to have `include/mspass/utility` and `include/mspass/seismic`. However, that would make the include lines unnecessarily verbose. I can see two solutions to this: 1) flatten the directory structure, and put all the headers under `include/mspass`; or 2) hide that from the user with a `include/mspass/mspass.h` file that includes all the public APIs. I think 2) is a cleaner solution. In addition to that public header file, we probably want to pull out the internal APIs from those header files. The apparent ones are the basic classes (`BasicMetadata.h`, `BasicTimeSeries.h`, etc.) should all be in a `include/internal` directory (probably `include/internal/utility` and `include/internal/seismic`) which will not be installed during a make install command.

**Comment by wangyinz** [comment]
I agree that we need to shift the focus to Python soon. I am not sure how should we design the database handle right now. I think we did not make it clear in our original design (refer to the mspass proposal). I think the most important thing is that we do want to take as much advantage as we could from ObsPy, so it is critical to decide what is the role of ObsPy to mspass. This would also affect whether we want to implement the database interface natively in Python. From my recent experiment with mongodb and obspy, it is pretty straightforward to make them work together, and saving a separate waveform data in GridFS is pretty sensible. Something that I don't know right now is how clean is mongodb's cxx API, and whether a Python-C binding will become awkward or partly incompatible. Anyway, the topic is getting pretty complex here, I would like to bring some of these discussions offline.

**Comment by wangyinz** [comment]
Checked by Codex

This looks like a good manual-close candidate. The current codebase has the C++/Python data object hierarchy in place, with `TimeSeries`, `Seismogram`, ensemble types, `ProcessingHistory`, and database-layer APIs exercised by the current test suite. The original architecture discussion appears superseded by the implemented design.

**Comment by wangyinz** [comment]
Closing this as completed/superseded by the implemented architecture.

The C++ data hierarchy is now established under `cxx/include/mspass/seismic/` (`TimeSeries`, `Seismogram`, and ensemble types), with `ProcessingHistory`/`ErrorLogger` in the utility layer. Database access is kept separate in the Python `mspasspy.db` layer rather than embedding a MongoDB handle in each C++ datum, which resolves the central design choice raised here.

Any new API change should be tracked as a focused issue against the current interfaces.
